src/ast/passes/resolve_imports.cpp

```diff
@@ -1,40 +1,326 @@
+                                        const std::string &name,
+                                        const std::string &contents)
+{
+  if (imports.scripts.contains(name)) {
+    return OK(); // Already added.
+  }
+
+  // Construct our context.
+  auto [it, added] = imports.scripts.emplace(name, ASTContext(name, contents));
+  assert(added);
+  auto &ast = it->second;
+
+  // Perform the basic parse pass. Note that this parse is done extremely
+  // early, and does zero expansion or parsing of attachpoints, etc.
+  PassManager pm;
+  pm.put(ast);
+  pm.put(bpftrace_);
+  pm.add(CreateParsePass());
+
+  // Attempt to parse the source.
+  auto ok = pm.run();
+  if (!ok) {
+    return ok.takeError();
+  }
+
+  // Disallow `config` blocks as they cannot be merged.
+  if (ast.root != nullptr && ast.root->config != nullptr &&
+      !ast.root->config->stmts.empty()) {
+    ast.root->config->addError() << "invalid `config` within import";
+  }
+
+  // Recursively visit the parsed tree.
+  visit(ast.root);
+
+  return OK();
+}
+
+Result<OK> ResolveImports::importObject([[maybe_unused]] Node &node,
+                                        const std::string &name,
+                                        const std::filesystem::path &path)
+{
+  if (imports.objects.contains(name)) {
+    return OK(); // Already added.
+  }
+
+  auto [_, added] = imports.objects.emplace(name, ExternalObject(path));
+  assert(added);
+  return OK();
+}
+
+Result<OK> ResolveImports::importBitcode([[maybe_unused]] Node &node,
+                                         const std::string &name,
+                                         const std::string &contents)
+{
+  if (imports.bitcode.contains(name)) {
+    return OK(); // Already added.
+  }
+
+  auto [_, added] = imports.bitcode.emplace(name, Bitcode(contents));
+  assert(added);
+  return OK();
+}
+
+Result<OK> ResolveImports::importAny(Node &node,
+                                     const std::string &name,
+                                     const std::filesystem::path &path,
+                                     bool ignore_unknown)
+{
+  if (!checkPerms(path)) {
+    node.addError()
+        << "cowardly refusing to import global writable permissions: " << path;
+    return OK();
+  }
+  if (std::filesystem::is_directory(path)) {
+    // Recursively import all entries in the directory. Note that the directory
+    // iterator will never include '.' or '..' entries.
+    for (const auto &entry : std::filesystem::directory_iterator(path)) {
+      auto ok = importAny(node,
+                          name + "/" + entry.path().filename().string(),
+                          entry.path(),
+                          true);
+      if (!ok) {
+        return ok.takeError();
+      }
+    }
+    return OK();
+  } else {
+    // Import any support file-based extensions.
+    if (path.extension() == ".bt") {
+      return importScript(node, name, path);
+    } else if (path.extension() == ".o") {
```

Comment:
Is it silly to be even more strict here and only look for something like `bpf.o` or `bt.o` ?

Reply:
Sure, I can do `bpf.o`. But `.bt` is the correct one, for bitcode files. It's not an object.

Reply:
Yeah let's maybe do `bpf.o`

Reply:
Done. Note that we need a lot more tests for this broadly, but I don't really want to load it all into this PR. I think it's reasonable to gate having imports be an "official" feature on this, I've really just added some minimal tests to ensure that the path resolution has some of the intended behaviors.